### ISSUE TITLE
[BUG]: matmul example crashes if the size is changed from 512x512 to 512x4096

### ISSUE THREAD

**Issue by dsharlet** [modularml/mojo#699]
### Bug description

I modified the matmul example to compute multiplications of 512x4096 x 4096x512. This works for the first few benchmarks, but crashes the tiled version:

```
munmap_chunk(): invalid pointer
[20205:20205:20230909,205744.348761:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq: No such file or directory (2)
[20205:20205:20230909,205744.348814:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq: No such file or directory (2)
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
Stack dump:
0.      Program arguments: mojo matmul.mojo
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  mojo      0x00005613e8ab3957
1  mojo      0x00005613e8ab152e
2  mojo      0x00005613e8ab402f
3  libc.so.6 0x00007f8c3119e520
4  libc.so.6 0x00007f8c311f2a7c pthread_kill + 300
5  libc.so.6 0x00007f8c3119e476 raise + 22
6  libc.so.6 0x00007f8c311847f3 abort + 211
7  libc.so.6 0x00007f8c311e56f6
8  libc.so.6 0x00007f8c311fcd7c
9  libc.so.6 0x00007f8c311fd05c
10 libc.so.6 0x00007f8c3120151a free + 186
11 libc.so.6 0x00007f8b7c005014 free + 18446744070670793652
Aborted
```

### Steps to reproduce

- git clone <mojo git url>
- Apply the following diff:
```@@ -239,11 +240,11 @@ fn main():
     print("Throughput of a 128x128 matrix multiplication in Python: ")
     let python_gflops = run_matmul_python(128, 128, 128)
     alias M = 512
+    alias N = 512
+    alias K = 4096
     # Mojo variants
     benchmark[matmul_naive](
-        M,
-        M,
-        M,
+        M, N, K,
         python_gflops,
         (
             "Throughput of a 512x512 matrix multiplication in Mojo using a"
@@ -251,9 +252,7 @@ fn main():
         ),
     )
     benchmark[matmul_vectorized_0](
-        M,
-        M,
-        M,
+        M, N, K,
         python_gflops,
         (
             "Throughput of a 512x512 matrix multiplication in Mojo using"
@@ -261,9 +260,7 @@ fn main():
         ),
     )
     benchmark[matmul_vectorized_1](
-        M,
-        M,
-        M,
+        M, N, K,
         python_gflops,
         (
             "Throughput of a 512x512 matrix multiplication in Mojo using the"
@@ -271,9 +268,7 @@ fn main():
         ),
     )
     benchmark[matmul_parallelized](
-        M,
-        M,
-        M,
+        M, N, K,
         python_gflops,
         (
             "Throughput of a 512x512 {vectorized + parallelized} matrix"
@@ -281,9 +276,7 @@ fn main():
         ),
     )
     benchmark[matmul_tiled_parallelized](
-        M,
-        M,
-        M,
+        M, N, K,
         python_gflops,
         (
             "Throughput of a 512x512 {tiled + vectorized + parallelized} matrix"
@@ -291,9 +284,7 @@ fn main():
         ),
     )
     benchmark[matmul_tiled_unrolled_parallelized](
-        M,
-        M,
-        M,
+        M, N, K,
         python_gflops,
         (
             "Throughput of a 512x512 {tiled + unrolled + vectorized +"
```
-  `mojo matmul.mojo`

### System information

```shell
- Ubuntu 22.04 on WSL2
- mojo 0.2.1 (64d14e85)
- modular 0.1.4 (6b54d308)
```

**Comment by dsharlet** [comment]
As an aside, this example really could use a test. It's incredibly easy to break the correctness of these functions, but the examples just assume the result is correct. I added this to support my hacking on this code:
```
        func(C, A, B, rt)
        var C_ref = Matrix(M, N)
        C_ref.zero()
        matmul_naive(C_ref, A, B, rt)
        for i in range(0, M):
            for j in range(0, N):
                if C[i, j] != C_ref[i, j]:
                    print("Mismatch: ", i, ", ", j, ": ", C[i, j], " != ", C_ref[i, j])
                    return
```
(Also needs A and B to be filled with non-zero data.)

**Comment by ematejska** [comment]
Might be duplicate of https://github.com/modularml/mojo/issues/586 but stack trace is different.

**Comment by jackos** [comment]
Thanks @dsharlet good idea, will likely add an assertion to restrict the dimensions that can be used

**Comment by dsharlet** [comment]
Perhaps I'm misunderstanding, I think failures where `M != N != K` are worth fixing, if anyone uses this as a basis for real code, they're going to have to debug this. For compiler/performance optimization work, large `K` with small `M` and `N` is useful for measuring the performance of the inner loop code quality (which is why I tried to make `K` big.

I looked a little closer at it and I think this is the fix, to be applied in the two tiling implementations:
```
-        tile[calc_tile, nelts * tile_size, tile_size](A.cols, C.cols)
+        tile[calc_tile, nelts * tile_size, tile_size](C.cols, B.rows)
```
Now it can work with `K` != `M` as long as `K` divides the split factor, and it also passes the correctness test I added in my local hacking branch.
